### PR TITLE
User Story 60501: 101-aks-cluster

### PR DESCRIPTION
**Pull request by TomArcherMsft**
Part of POC to test generating sample code and articles using OpenAI.

### CODE REVIEW

File: quickstart/101-aks-cluster/main.tf [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+resource "azapi_resource" "ssh_public_key" {
+  type      = "Microsoft.Compute/sshPublicKeys@2022-11-01"
+  name      = random_pet.ssh_key_name.id
+  location  = "westus3"
+  parent_id = azurerm_resource_group.rg.id
+}
+
+resource "azapi_resource_action" "ssh_public_key_gen" {
```
Comment (lonegunmanb) [inline quickstart/101-aks-cluster/main.tf]: I would recommend [`tls_private_key`](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/resources/private_key) resource:

```hcl
# RSA key of size 4096 bits
resource "tls_private_key" "rsa_4096" {
  algorithm = "RSA"
  rsa_bits  = 4096
}
```

File: quickstart/101-aks-cluster/main.tf [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    node_count      = var.agent_count
+    vm_size         = var.agent_vm_size
+    os_disk_size_gb = var.os_disk_size_gb
+  }
+
+  linux_profile {
+    admin_username = var.linux_admin_username
+
+    ssh_key {
+      key_data = jsondecode(azapi_resource_action.ssh_public_key_gen.output)["publicKey"]
```
Comment (lonegunmanb) [inline quickstart/101-aks-cluster/main.tf:39]: If we use `tls_private_key` resource, then the `key_data` could be:

```hcl
key_data = tls_private_key.rsa_4096.public_key_openssh
```

File: quickstart/101-aks-cluster/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,20 @@
+output "resource_group_name" {
+  value = azurerm_resource_group.rg.name
+}
+
+output "ssh_key_name" {
```
Comment (lonegunmanb) [inline quickstart/101-aks-cluster/outputs.tf:5]: We can export the private key:

```hcl
output "ssh_private_key_openssh" {
  sensitive = true
  value = tls_private_key.rsa_4096.private_key_openssh
}

output "ssh_private_key_pem" {
  sensitive = true
  value = tls_private_key.rsa_4096.private_key_pem
}
```

File: quickstart/101-aks-cluster/providers.tf [hunk 4]
```diff
@@ -0,0 +1,20 @@
+terraform {
+  required_version = ">=1.0"
+  required_providers {
+    azapi = {
```
Comment (lonegunmanb) [inline quickstart/101-aks-cluster/providers.tf:4]: We can use `tls` provider here:

```hcl
tls = {
      source = "hashicorp/tls"
      version = "~>4.0"
    }
```

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @TomArcherMsft for opening this pr, could we use `tls_private_key` resource to generate the ssh key?

File: quickstart/101-aks-cluster/main.tf (hunk 1 above)
Comment (TomArcherMsft) [reply quickstart/101-aks-cluster/main.tf]: Hi, @lonegunmanb. I used the AzAPI as that is something @grayzu recommended in an email thread with all of us. Maybe we need to reengage on the email thread or figure it out here?

**Review by lonegunmanb — APPROVED** [review]
LGTM! @grayzu WDYT?

File: quickstart/101-aks-cluster/main.tf (hunk 1 above)
Comment (grayzu) [reply quickstart/101-aks-cluster/main.tf]: Although the tls provider will do the trick, I think making use of the Azure functionality which will provide SSH certificates that can be used in production environments is a better way to show this functionality. According to the docs, the tls provider is not recommended for prod use.

**Review by grayzu — APPROVED** [review]
I responded to lonegunmanb's suggestion to use tls provider. Everything else looks great.

**Comment by TomArcherMsft** [comment]
@grayzu @lonegunmanb 

In a more recent [PR](https://github.com/Azure/terraform/pull/225/), I put the SSH-creation code in a [separate file (ssh.tf)](https://github.com/Azure/terraform/pull/225/files#diff-785525a1ce061b4466ebf940782c071fe5515e9e156a1dd885376ef3d827448e). I like doing that as it's isolates the SSH creation & AzAPI usage and reduces code in main.tf. Should we use that pattern as the standard? If so, I'll make the change to this PR.

**Comment by lonegunmanb** [comment]
> @grayzu @lonegunmanb
> 
> In a more recent [PR](https://github.com/Azure/terraform/pull/225/), I put the SSH-creation code in a [separate file (ssh.tf)](https://github.com/Azure/terraform/pull/225/files#diff-785525a1ce061b4466ebf940782c071fe5515e9e156a1dd885376ef3d827448e). I like doing that as it's isolates the SSH creation & AzAPI usage and reduces code in main.tf. Should we use that pattern as the standard? If so, I'll make the change to this PR.

As long as we've agreed on a working pattern I think it's ok to apply the pattern on this repo.

**Comment by TomArcherMsft** [comment]
@stemaMSFT @lonegunmanb This PR is ready to review.

**Review by Jingwei-MS — APPROVED** [review]
LGTM~

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Hi @TomArcherMsft thanks for the update, it looks this pr's base commit is too old so the changed files scope that the pipeline calculated was wrong. Would you please rebase your branch to the latest master branch and try again? Thanks!

**Pull request closed without merging**